File: src/Physics/aleMeshHydroHLLE.cc
// Copyright (C) 2026  Cody Raskin

#pragma once
#include "aleMeshHydroBase.hh"
#include "HLL.cc"

// Two-wave HLLE on ALEMesh -- more diffusive than ALEMeshHydroHLLC (no exact
// contact resolution), but without HLLC's star-region solve, so it has no
// equivalent of the sonic-point entropy-violation weakness that limits HLLC
// to a short stable window on a strong discontinuity. First-order (no
// reconstruction), same as ALEMeshHydroHLLC -- see aleMeshHydroBase.hh.
template<int dim>
class ALEMeshHydroHLLE : public ALEMeshHydroBase<dim> {
public:
    using typename ALEMeshHydroBase<dim>::Vector;
    using typename ALEMeshHydroBase<dim>::VectorField;
    using typename ALEMeshHydroBase<dim>::ScalarField;

    ALEMeshHydroHLLE(NodeList* nodeList, PhysicalConstants& constants,
                      EquationOfState* eos, Mesh::ALEMesh<dim>* mesh,
                      NodeList* nodeVelocities = nullptr)
        : ALEMeshHydroBase<dim>(nodeList, constants, eos, mesh, nodeVelocities) {}

    virtual std::string name() const override { return "ALEMeshHydroHLLE"; }

    virtual HLLFlux<dim>
    computeFlux(double rhoL, const Vector& vL, double uL, double pL, double csL,
                double rhoR, const Vector& vR, double uR, double pR, double csR,
                const Vector& normal, double faceVelocity) const override {
        return computeHLLEFluxFromStatesALE<dim>(rhoL, vL, uL, pL, csL,
                                                  rhoR, vR, uR, pR, csR,
                                                  normal, faceVelocity);
    }
};

File: src/Physics/aleMeshHydroHLLE.py
# Copyright (C) 2026  Cody Raskin

from PYB11Generator import *
from hydro import *

@PYB11template("dim")
class ALEMeshHydroHLLE(Hydro):
    def pyinit(self,
               nodeList="NodeList*",
               constants="PhysicalConstants&",
               eos="EquationOfState*",
               mesh="Mesh::ALEMesh<%(dim)s>*",
               nodeVelocities=("NodeList*", "nullptr")):
        "nodeVelocities: optional node-centered NodeList (one entry per mesh node) with a Vector 'velocity' field prescribing mesh motion. Omit (or pass None) for a static (Eulerian) mesh."
        return

ALEMeshHydroHLLE2d = PYB11TemplateClass(ALEMeshHydroHLLE,
                              template_parameters = ("2"),
                              cppname = "ALEMeshHydroHLLE<2>",
                              pyname = "ALEMeshHydroHLLE2d",
                              docext = " (2D).")
